### DISCUSSION TITLE
How to combine all same node by using networkx

### DISCUSSION THREAD

**Discussion by bigpapame** [networkx/networkx#4606] [Q&A]
How to combine specific node by using networkx ?

Here is part of my gml file
First thing is to read the gml file into networkx,
after reading the file, i want to combine all 'shu.edu.tw'  into one node
For instant：id 0 and id 4 have the same 
'shu.edu.tw' , i want to combine id 4 into id 0, same thing does the id 5 and 6, and all the edges have to make up to the previous one. 

``` 
graph [
  directed 1
  multigraph 1
  node [
    id 0
    label "https://www.shu.edu.tw/"
  ]
  node [
    id 1
    label "https://www.facebook.com"
  ]
  node [
    id 2
    label "https://www.instagram.com"
  ]
  node [
    id 3
    label "https://www.youtube.com"
  ]
  node [
    id 4
    label "http://cc.web.shu.edu.tw"
  ]
  node [
    id 5
    label "http://donate.web.shu.edu.tw"
  ]
  node [
    id 6
    label "http://english.shu.edu.tw"
  ]
```

**Comment by mrecachinas** [comment]
@bigpapame Take a look at [nx.contracted_nodes](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.minors.contracted_nodes.html).

If I'm understanding correctly, you want to combine nodes (while still preserving edges) with ids 0, 4, 5, and 6 (specifically, based on having `shu.edu.tw` in the domain name). If that's your goal, you'd probably do something like:

```python
G = nx.read_gml(...)
print(G.nodes)
# => NodeView(('https://www.shu.edu.tw/', 'https://www.facebook.com', 'https://www.instagram.com', 'https://www.youtube.com', 'http://cc.web.shu.edu.tw', 'http://donate.web.shu.edu.tw', 'http://english.shu.edu.tw'))

# note: this will merge the nodes such that only "https://www.shu.edu.tw/" will remain
# also note: copy=False will modify the graph in-place
nx.contracted_nodes(G, "https://www.shu.edu.tw/", "http://cc.web.shu.edu.tw", copy=False)
nx.contracted_nodes(G, "https://www.shu.edu.tw/", "http://donate.web.shu.edu.tw", copy=False)
nx.contracted_nodes(G, "https://www.shu.edu.tw/", "http://english.shu.edu.tw", copy=False)

print(G.nodes)
# => NodeView(('https://www.shu.edu.tw/', 'https://www.facebook.com', 'https://www.instagram.com', 'https://www.youtube.com'))
```

If you wanted to do it based on checking the label, you'd just need to loop over `G.nodes`, something like:

```python
G = nx.read_gml(...)
print(G.nodes)
# => NodeView(('https://www.shu.edu.tw/', 'https://www.facebook.com', 'https://www.instagram.com', 'https://www.youtube.com', 'http://cc.web.shu.edu.tw', 'http://donate.web.shu.edu.tw', 'http://english.shu.edu.tw'))

# copy the list of nodes, since we're modifying
# G.nodes as we iterate
nodes = list(G.nodes)
for u in nodes:
    for v in nodes:
        if u != v and G.has_node(u) and G.has_node(v) and "shu.edu.tw" in u and "shu.edu.tw" in v:
            nx.contracted_nodes(G, u, v, copy=False)
print(G.nodes)
# => NodeView(('https://www.shu.edu.tw/', 'https://www.facebook.com', 'https://www.instagram.com', 'https://www.youtube.com'))
```

Just to explain the loop a bit: The `u != v` is because we don't want to waste time combining a node with itself. The `G.has_node(u)` and `G.has_node(v)` is because we're modifying the graph as we're looping through. Since we aren't updating our extracted `nodes` list, we want to make sure to skip already combined nodes. The `"shu.edu.tw" in u` and `"shu.edu.tw" in v` is to combine nodes based on a common domain name.

Note: `nx.contracted_nodes` also has a kwarg `self_loop=True/False` for keeping or getting rid of self-loops (i.e., by default, edges between `u` and `v` with turn into self-loops).

**Comment by bigpapame** [comment]
Will ```nx.contracted``` nodes merge the weight of the subsequent edge into the original edge?

**Comment by bigpapame** [comment]
After running this code, i got this error 
```TypeError: contracted_nodes() got an unexpected keyword argument 'copy' ```

**Comment by mrecachinas** [comment]
What version of `networkx` are you using? It looks like the `copy` keyword argument was added in 2.5. If you're using an older version of networkx, you can simply assign the returned contracted graph to a variable, like the following:

```python
G = nx.contracted_nodes(G, u, v)
```

**Comment by mrecachinas** [comment]
`nx.contracted_nodes` will merge the weights and edges for an `nx.Graph` (I believe the last contracted node will "win" -- i.e., override the edge weight); however, I think `nx.contracted_nodes` _will_ maintain all edges and associated weights for a `nx.MultiGraph`.
